pkg/yurtiotdock/clients/edgex-foundry/util.go

```diff
@@ -0,0 +1,456 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      toKubeName(ds.Name),
+			Namespace: "default",
+			Labels: map[string]string{
+				EdgeXObjectName: ds.Name,
+			},
+		},
+		Spec: iotv1alpha1.DeviceServiceSpec{
+			Description: ds.Description,
+			Labels:      ds.Labels,
+			AdminState:  iotv1alpha1.AdminState(ds.AdminState),
+			BaseAddress: ds.BaseAddress,
+		},
+		Status: iotv1alpha1.DeviceServiceStatus{
+			EdgeId:        ds.Id,
+			LastConnected: ds.LastConnected,
+			LastReported:  ds.LastReported,
+			AdminState:    iotv1alpha1.AdminState(ds.AdminState),
+		},
+	}
+}
+
+func toEdgeXDevice(d *iotv1alpha1.Device) dtos.Device {
+	md := dtos.Device{
+		Description:    d.Spec.Description,
+		Name:           getEdgeXName(d),
+		AdminState:     string(toEdgeXAdminState(d.Spec.AdminState)),
+		OperatingState: string(toEdgeXOperatingState(d.Spec.OperatingState)),
+		Protocols:      toEdgeXProtocols(d.Spec.Protocols),
+		LastConnected:  d.Status.LastConnected,
+		LastReported:   d.Status.LastReported,
+		Labels:         d.Spec.Labels,
+		Location:       d.Spec.Location,
+		ServiceName:    d.Spec.Service,
+		ProfileName:    d.Spec.Profile,
+	}
+	if d.Status.EdgeId != "" {
+		md.Id = d.Status.EdgeId
+	}
+	return md
+}
+
+func toEdgeXUpdateDevice(d *iotv1alpha1.Device) dtos.UpdateDevice {
+	adminState := string(toEdgeXAdminState(d.Spec.AdminState))
+	operationState := string(toEdgeXOperatingState(d.Spec.OperatingState))
+	md := dtos.UpdateDevice{
+		Description:    &d.Spec.Description,
+		Name:           &d.Name,
+		AdminState:     &adminState,
+		OperatingState: &operationState,
+		Protocols:      toEdgeXProtocols(d.Spec.Protocols),
+		LastConnected:  &d.Status.LastConnected,
+		LastReported:   &d.Status.LastReported,
+		Labels:         d.Spec.Labels,
+		Location:       d.Spec.Location,
+		ServiceName:    &d.Spec.Service,
+		ProfileName:    &d.Spec.Profile,
+		Notify:         &d.Spec.Notify,
+	}
+	if d.Status.EdgeId != "" {
+		md.Id = &d.Status.EdgeId
+	}
+	return md
+}
+
+func toEdgeXProtocols(
+	pps map[string]iotv1alpha1.ProtocolProperties) map[string]dtos.ProtocolProperties {
+	ret := map[string]dtos.ProtocolProperties{}
+	for k, v := range pps {
+		ret[k] = dtos.ProtocolProperties(v)
+	}
+	return ret
+}
+
+func toEdgeXAdminState(as iotv1alpha1.AdminState) models.AdminState {
+	if as == iotv1alpha1.Locked {
+		return models.Locked
+	}
+	return models.Unlocked
+}
+
+func toEdgeXOperatingState(os iotv1alpha1.OperatingState) models.OperatingState {
+	if os == iotv1alpha1.Up {
+		return models.Up
+	} else if os == iotv1alpha1.Down {
+		return models.Down
+	}
+	return models.Unknown
+}
+
+// toKubeDevice serialize the EdgeX Device to the corresponding Kubernetes Device
+func toKubeDevice(ed dtos.Device) iotv1alpha1.Device {
+	var loc string
+	if ed.Location != nil {
+		loc = ed.Location.(string)
+	}
+	return iotv1alpha1.Device{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      toKubeName(ed.Name),
+			Namespace: "default",
```

Comment:
Would it be more appropriate to set the namespace here to match that of yurt-iot-dock?